test/integration/statefulset/statefulset_test.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statefulset
+
+import (
+	"fmt"
+	"testing"
+
+	"k8s.io/api/apps/v1beta1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestSpecReplicasChange(t *testing.T) {
+	s, closeFn, rm, informers, c := scSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-spec-replicas-change", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	stopCh := runControllerAndInformers(rm, informers)
+	defer close(stopCh)
+
+	createHeadlessService(t, c, newHeadlessService(ns.Name))
+	sts := newSTS("sts", ns.Name, 2)
+	stss, _ := createSTSsPods(t, c, []*v1beta1.StatefulSet{sts}, []*v1.Pod{})
+	sts = stss[0]
+	waitSTSStable(t, c, sts)
+
+	// Update .Spec.Replicas and verify .Status.Replicas is changed accordingly
+	scaleSTS(t, c, sts, 3)
+	scaleSTS(t, c, sts, 0)
+	scaleSTS(t, c, sts, 2)
+
+	// Add a template annotation change to test STS's status does update
+	// without .Spec.Replicas change
+	stsClient := c.AppsV1beta1().StatefulSets(ns.Name)
+	var oldGeneration int64
+	newSTS := updateSTS(t, stsClient, sts.Name, func(sts *v1beta1.StatefulSet) {
+		oldGeneration = sts.Generation
+		sts.Spec.Template.Annotations = map[string]string{"test": "annotation"}
+	})
+	savedGeneration := newSTS.Generation
+	if savedGeneration == oldGeneration {
+		t.Fatalf("Failed to verify .Generation has incremented for sts %s", sts.Name)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
```

Comment:
Polling will be used a lot for controllers like StatefulSet, so it may be wise to make this a helper function (something like [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L912-L923)).